# Process data

Thoughts

Tables, attributes that are important

Requirements are in requirements.txt

In [24]:
import pandas as pd

Sheets: ['AMX_OFFER_HEADER', 'AMX_OFFER_LINES', 'AMX_DONATION_HEADER', 'AMX_DONATION_LINES', 'ACBIDS_ARCHIVE', 'Choice - DOCUMENTHEADER', 'Choice - DOCUMENTLINES', 'RW_ORG', 'RW_USER']

Corresponds to the data dictionary table name

In [ ]:
sheets = pd.ExcelFile('./data/Choice Data.xlsx').sheet_names
print(sheets)

In [ ]:
data = pd.ExcelFile('./data/Choice Data.xlsx')

tables = {}

for table in data.sheet_names:
    df = pd.read_excel(data, sheet_name=table, nrows=0)
    tables[table] = df.columns.to_list()
    
# print(tables)

print('len:', len(tables['AMX_DONATION_HEADER']))

'AMX_OFFER_HEADER': ['OFFERNUMBER', 'ANAME', 'ACONTACT', 'ACITY', 'ASTATE', 'APHONE', 'OFFERDATE', 'OFFERTIME', 'OFFEREXPIRATIONDATE', 'OFFEREXPIRATIONTIME', 'OUTBYDATE', 'FIRSTHALFHOURFREE', 'DONORID', 'DONORNAME', 'DONORREFERENCENUMBER', 'RESTRICTIONS', 'WCONTACTNAME', 'WAREHOUSENAME', 'WCONTACTADDRESS', 'WCONTACTCITY', 'WCONTACTSTATE', 'WCONTACTZIP', 'RELEASEDATE', 'SHIPMENTMETHOD', 'STAGINGMETHOD', 'RELEASENUMBER', 'CREDITEXPIRATIONTIME', 'LOGISTICSREP', 'RECEIPTDUEDATE', 'LOGREPPHONE', 'LOGREPFAX', 'LOGREPEMAIL', 'MILES', 'MAGREFNUMBER', 'DONATIONNUMBER', 'STATUS', 'HTMLFILENAME', 'DATEPICKEDUP', 'DECLINEREASONCODE', 'WCONTACTPHONE', 'WCONTACTEMAIL', 'RECEIPTEDBY', 'AFFILIATECOMMENTS', 'TRANSPORTATION', 'DELIVERYINSTRUCTIONS', 'LOADNUMBER'],

'AMX_OFFER_LINES': ['MAGREFNUMBER', 'ITEMDESCRIPTION', 'PRODUCTTYPECODE', 'PACK', 'SIZE_', 'SIZEUOM', 'QUANTITY', 'QUANTITYPERPALLET', 'GROSSWEIGHT', 'NETWEIGHT', 'CUBE', 'MANUFACTUREDATE', 'SHELFLIFE', 'EXTENSION', 'UNITOFMEASURE', 'TOTALPALLETS', 'TOTALGROSSWEIGHT', 'TOTALCUBE', 'STORAGEREQUIREMENT', 'EXPIRATIONDATE', 'USEBYDATE', 'BILLOFLADINGNUMBER', 'PACKAGINGTYPE', 'DONATIONREASON', 'ITEMNUMBER', 'OFFERNUMBER', 'OUTBYDATE', 'DONORITEMNO', 'QUANTITYRECEIVED', 'PALLETSRECEIVED', 'GROSSWEIGHTRECEIVED', 'GROUPID'],

'AMX_DONATION_HEADER': ['DONATIONNUMBER', 'DONORID', 'DONORNAME', 'DCONTACTNAME', 'DCONTACTCITYSTATEZIP', 'DCONTACTPHONE', 'DCONTACTFAX', 'DONORREFERENCENUMBER', 'DONATIONDATE', 'RELEASEDATE', 'DCONTACTEMAIL', 'RCONTACTNAME', 'RCONTACTPHONE', 'RCONTACTEMAIL', 'WCONTACTNAME', 'WAREHOUSENAME', 'WCONTACTADDRESS', 'WCONTACTCITYSTATEZIP', 'WCONTACTPHONE', 'WCONTACTFAX', 'RELEASENUMBER', 'SHIPMENTMETHOD', 'STAGINGMETHOD', 'DELIVERYINSTRUCTIONS', 'STATUS', 'MAGREFNUMBER', 'HTMLFILENAME', 'OUTBYDATE', 'RCONTACTFAX', 'WCONTACTEMAIL'], 

'AMX_DONATION_LINES': ['DONATIONNUMBER', 'ITEMNUMBER', 'ITEMDESCRIPTION', 'DONATIONREASON', 'PACK', 'SIZE_', 'SIZEUOM', 'PACKAGINGTYPE', 'UNITOFMEASURE', 'QUANTITY', 'QUANTITYPERPALLET', 'TOTALPALLETS', 'GROSSWEIGHT', 'TOTALGROSSWEIGHT', 'NETWEIGHT', 'CUBE', 'STORAGEREQUIREMENT', 'MANUFACTUREDATE', 'EXPIRATIONDATE', 'SHELFLIFE', 'EXTENSION', 'USEBYDATE', 'BILLOFLADINGNUMBER', 'STATUS', 'CODEDATETYPE', 'MAGREFNUMBER'], 

'ACBIDS_ARCHIVE': ['DOCUMENTID', 'AFFILIATEWEBID', 'BIDAMOUNT', 'BIDDATE', 'BIDTIME', 'RANK', 'AVAILABLELOADS', 'STATUS', 'ISINITIATOR', 'GROUPID', 'PASSKEY', 'LOADS', 'REASON', 'WINNINGBID', 'USERNAME', 'CREDITSHARES', 'SHARES', 'GROSSWEIGHT', 'DESCRIPTION', 'ID', 'INITIATORWEBID', 'WONLOAD', 'WINNINGBIDLOAD', 'AUCTIONDATE', 'SESSIONINFO', 'TRANSPORTATION'], 

'Choice - DOCUMENTHEADER': ['DOCUMENTID', 'REFNUMBER', 'SOURCEORG', 'USERID', 'STATUS', 'CREATEDDATE', 'EXPIRYDATE', 'MODIFIEDDATE', 'WAREHOUSEID', 'PROP1', 'PROP2', 'PROP3', 'PROP4', 'PROP5', 'PROP6', 'PROP7', 'PROP8', 'PROP9', 'PROP10', 'PROP11', 'PROP12', 'PROP13', 'PROP14', 'PROP15', 'NOTIFY', 'DOCTYPE', 'USERNAME', 'USEREMAIL', 'SOURCEORGNAME', 'DOMAINID', 'DOMAINNAME', 'PROP16', 'PROP17', 'PROP18', 'PROP19', 'PROP20', 'PROP21', 'PROP22', 'PROP23', 'PROP24', 'PROP25', 'PROP26', 'PROP27', 'PROP28', 'PROP29', 'PROP30', 'PROP31', 'PROP32', 'PROP33', 'PROP34', 'PROP35'], 

'Choice - DOCUMENTLINES': ['HEADERID', 'LINEITEMNUMBER', 'ITEMNAME', 'ITEMCODE', 'TRADINGGROUP', 'DESTINATIONORG', 'DONATEDQUANTITY', 'PROP1', 'PROP2', 'PROP3', 'PROP4', 'PROP5', 'PROP6', 'PROP7', 'PROP8', 'PROP9', 'PROP10', 'PROP11', 'PROP12', 'PROP13', 'PROP14', 'PROP16', 'PROP17', 'PROP18', 'PROP19', 'PROP20', 'NOTIFY', 'PROP15', 'STATUS', 'ACCEPTEDQUANTITY', 'ACCEPTEDDATE', 'RECEIPTEDQUANTITY', 'RECEIPTEDDATE', 'DESTINATIONORGNAME', 'COMMENTS', 'USERID', 'USERFIRSTNAME', 'USERLASTNAME', 'USEREMAIL', 'RPROP1', 'RPROP2', 'RPROP3', 'RPROP4', 'RPROP5', 'RPROP6', 'RPROP7', 'RPROP8', 'RPROP9', 'RPROP10', 'RPROP11', 'RPROP12', 'RPROP13', 'RPROP14', 'RPROP15', 'PROP21', 'PROP22', 'PROP23', 'PROP24', 'PROP25', 'PROP26', 'PROP27', 'PROP28', 'PROP29', 'PROP30'], 

'RW_ORG': ['[', 'NAME', 'PRINTABLE_NAME', 'STATUS_FLAG', 'E_MAIL', 'DESCRIPTION', 'ONE_APPROVER_FLAG', 'FINAL_AMOUNT', 'CURRENCY', 'CURRENCY_PATTERN', 'FLOAT_PATTERN', 'GROUP_SEPARATOR', 'GROUP_SIZE', 'DECIMAL_CHAR', 'LANGUAGE', 'COUNTRY', 'VARIANT', 'TIME_ZONE', 'DOMAIN', 'URL', 'DIVISION_ID', 'EXTERNAL_ID', 'DUNS_NUMBER', 'TAX_ID', 'INDUSTRY_CODE', 'CONTACT_NAME', 'ORG_TYPE', 'CUSTOM_TYPE', 'CREATED_TIME', 'CREATED_BY', 'MODIFIED_TIME', 'MODIFIED_BY', 'SYS_FLAGS', 'PATH', 'YEARS_IN_BUSINESS', 'INDUSTRY_VAL_NUM', 'ANNUAL_REV_VID', 'EMP_COUNT_VID', 'OCA', 'PARENT_ORG_ID', 'FINAL_APPROVER_ID', 'ADMIN_USER_ID', 'SUPPLIER_ID', 'MASTER_ORG_ID'], 

'RW_USER': ['USER_ID', 'LOGIN_STR', 'STATUS_CODE', 'E_MAIL', 'LAST_NAME', 'FIRST_NAME', 'MIDDLE_NAME', 'PRINTABLE_NAME', 'EMPLOYEE_ID', 'TELEPHONE', 'FAX', 'TITLE', 'SYS_FLAGS', 'ADDRESS_LINE_1', 'ADDRESS_LINE_2', 'ADDRESS_LINE_3', 'CITY', 'STATE', 'COUNTRY', 'ZIP', 'SALUTATION_VID', 'IS_CONTACT_ALLOWED', 'PASSWORD_SET_DATE', 'PASSWORD_REMINDER', 'PASSWORD_QUESTION', 'PASSWORD_ANSWER', 'CREATED_BY', 'CREATED_TIME', 'MODIFIED_BY', 'MODIFIED_TIME', 'EXTERNAL_ID', 'OCA', 'ORG_ID', 'REPORTS_TO_ID', 'JOB_LEVEL_ID', 'JOB_FUNCTION_ID']

Rename RW_ORG column '[' to 'ORG_ID'

Relavant data/attributes

From choice data, check that

Posting Date: AMX_DONATION_HEADER --> DONATIONDATE

Donor ID: RW_ORG --> NAME

Gross Weight: AMX_DONATION_LINES --> TOTALGROSSWEIGHT

Branch Code: AMX_DONATION_HEADER --> WAREHOUSENAME

Storage Code: AMX_DONATION_LINES --> STORAGEREQUIREMENT

~~Donor Class of Trade: RW_ORG --> ORG_TYPE (Note that all ORG_TYPE are the same, which is noted "S")~~

Item Number: AMX_DONATION_LINES --> DONATIONNUMBER

We will mostly focus on the following sheets of the dataset "Choice Data.xlsx": 
    AMX_DONATION_LINES: records the donated items
    AMX_DONATION_HEADER: records the donors of each donated items and the warehouse info
    RW_ORG: records general info of all different donors

The three datasets are linked in the following way:
    Each donation has a DONATIONNUMBER. The DONATIONNUMBER is present in AMX_DONATION_LINES and AMX_DONATION_HEADER, which can link these two datasets, providing more information on each donated items. 
    Each donor has a DONORID. The DONORID is present in AMX_DONATION_HEADER and RW_ORG, which can link these two datasets, providing more information on donors.

Review sample of data.

xlsx > sqlite database!
Multipe tables/sheets/relations

https://emory-my.sharepoint.com/shared?listurl=https%3A%2F%2Femory%2Esharepoint%2Ecom%2Fsites%2FEmoryCenterforAILearningStaff%2FShared%20Documents&id=%2Fsites%2FEmoryCenterforAILearningStaff%2FShared%20Documents%2FGeneral%2FAI%2EData%20Lab%2FSpring%2025%2FProject%20Backgrounds%2FTechBridge%2FHunger%20Relief%20Product%20Info%2FEntity%20Relationship%20Diagrams%20%2D%20TechBridge%20%2D%20AI%2EDataLab%2Epdf&parent=%2Fsites%2FEmoryCenterforAILearningStaff%2FShared%20Documents%2FGeneral%2FAI%2EData%20Lab%2FSpring%2025%2FProject%20Backgrounds%2FTechBridge%2FHunger%20Relief%20Product%20Info

In [ ]:
import sqlite3

'''conn = sqlite3.connect("data/choice.db")

for sheet_name in data.sheet_names:
    df = pd.read_excel(data, sheet_name=sheet_name)
    df.to_sql(sheet_name, conn, index=False, if_exists="replace")

conn.commit()
conn.close()'''

# Rename column in RW_ORG from '[' to 'ORG_ID'

conn = sqlite3.connect('data/choice.db')
cursor = conn.cursor()

rename = """
ALTER TABLE rw_org
RENAME COLUMN '[' TO org_id;
"""

cursor.execute(rename)

conn.commit()
conn.close()

Queries

In [ ]:
conn = sqlite3.connect('data/choice.db')

query = 'SELECT name FROM sqlite_master WHERE type='table';"
# query = "SELECT * FROM AMX_OFFER_HEADER LIMIT 10;" # demo

# query = "SELECT * FROM amx_donation_header ORDER BY donationDate LIMIT 10;"

result = pd.read_sql_query(query, conn)

print(result)

conn.close()

                      name
0         AMX_OFFER_HEADER
1          AMX_OFFER_LINES
2      AMX_DONATION_HEADER
3       AMX_DONATION_LINES
4           ACBIDS_ARCHIVE
5  Choice - DOCUMENTHEADER
6   Choice - DOCUMENTLINES
7                   RW_ORG
8                  RW_USER
9               prediction


Translate Choice to time series compatible data.
Figure this out.

ORDER BY DONATIONDATE

Figure out how many tuples in each table.

In [18]:
conn = sqlite3.connect('data/choice.db')
cursor = conn.cursor()

get_table_names = "SELECT name FROM sqlite_master WHERE type='table';"
cursor.execute(get_table_names)
table_names = cursor.fetchall()

lengths = {}

for table_name in table_names:
    get_lengths = f"""
    SELECT COUNT(*)
    FROM '{table_name[0]}'
    """
    
    cursor.execute(get_lengths)
    lengths[table_name[0]] = cursor.fetchone()[0]

print(lengths)


{'AMX_OFFER_HEADER': 46509, 'AMX_OFFER_LINES': 359978, 'AMX_DONATION_HEADER': 13469, 'AMX_DONATION_LINES': 406616, 'ACBIDS_ARCHIVE': 5099, 'Choice - DOCUMENTHEADER': 11828, 'Choice - DOCUMENTLINES': 302671, 'RW_ORG': 80, 'RW_USER': 1923}


'AMX_OFFER_HEADER': 46509, 

'AMX_OFFER_LINES': 359978,

'AMX_DONATION_HEADER': 13469,

'AMX_DONATION_LINES': 406616,

'ACBIDS_ARCHIVE': 5099,

'Choice - DOCUMENTHEADER': 11828,

'Choice - DOCUMENTLINES': 302671,

'RW_ORG': 80, 

'RW_USER': 1923

In [ ]:
conn = sqlite3.connect('data/choice.db')
cursor = conn.cursor()

query = """
CREATE TABLE prediction AS
SELECT header.donationdate posting_date,
header.donorname donor_id,
lines.totalgrossweight gross_weight,
header.warehousename branch_code,
lines.storagerequirement storage_code
FROM amx_donation_header header
JOIN amx_donation_lines lines ON header.donationnumber = lines.donationnumber
"""

# Exclude rw_org since all 80 tuples are S org type.
# How to deal with null names?

cursor.execute(query)
print(cursor.fetchone())

conn.commit()
conn.close()

Check for similarity of columns to article data

Add more variables that we have available v. the paper?

In [ ]:
conn = sqlite3.connect('data/choice.db')
cursor = conn.cursor()

check = """
SELECT COUNT(*)
FROM prediction
GROUP BY donor_id
"""

cursor.execute(check)
print(len(cursor.fetchall()))

# 659 branches :( not good
# 7 storage codes :)

189
